packages/state-utils/src/keyed-reducer/index.ts

```diff
@@ -98,24 +123,33 @@ export const keyedReducer = ( keyPath, reducer ) => {
 					}
 					return result;
 				},
-				undefined
+				undefined as SerializationResult< TState > | undefined
```

Comment:
The result of serialization of `TState` doesn't need to be of type `TState`. It's a generic JSON-serializable value that can be then saved or loaded from local storage. What's the best TypeScript type to describe such a JSON value that usually comes from external sources like local storage or REST response? `JSON.parse` has a return type of `any`. Is there anything better?

A typical usage where the serialization/deserialization is not trivial identity is serializing class instances:
```js
const reducer: Reducer< QueryManager >( state: QueryManager = new QueryManager(), action ) => {
  switch ( action.type ) {
    case 'SERIALIZE':
      return state.toJSON();
    case 'DESERIALIZE':
      return QueryManager.fromJSON( state );
    case 'ADD_RESULTS_PAGE':
      return state.addResultsPage( action.results );
  }
} );
```

The serialization and deserialization with reducers introduces a lot of trouble into our types, because it reuses the reducer for three distinct purposes, where the types are different:
- `reducer( serialized, { type: 'DESERIALIZE' } )` receives value of type `JSON` and returns `TState`
- `reducer( state, { type: 'SERIALIZE' } )` receives `TState` and returns `JSON | undefined`
- `reducer( state, { type: 'FOO' } )` is the actual Redux usage, receiving and returning `TState`. And Redux `TState` is never `undefined`, unlike the result of serialization

Maybe making the `SERIALIZE` and `DESERIALIZE` action quack like Redux action although they really aren't Redux was a bad idea that's not compatible with neat types.

A better reimplementation of persistence would be additional methods on the reducer:
```js
const reducer: Reducer< TState > = ( state, action ) => {
  // pure Redux stuff only
};
reducer.serialize = ( state: TState ) => {
  return ser( state ) as JSON;
};
reducer.deserialize = ( json: JSON ) => {
  return deser( state ) as TState;
}
```

Refactoring the current codebase this way shouldn't be hard and would make types much easier. @sgomes what do you think?

Reply:
In my experience with serialisation code, I'm reluctant to accept any sentence that includes the words "shouldn't be hard", but the refactor sounds like an excellent idea! 👍 It's much more explicit about intention and result, and it could probably help us get rid of some confusing code.

TypeScript is great for finding inconsistencies like this; if something's hard to type, you may want to rethink the approach.

Reply:
What would be the best way to implement that new API, @jsnajdr? A new function, `withPersistence` that accepts a reducer function, a serialize function and a deserialize function?

```ts
type SerializableReducer< TState, TAction extends Action = AnyAction > = {
    ( state: TState, action: TAction ) => TState;
    serialize?: ( state: TState ) => JSON;
    deserialize?: ( json: JSON ) => TState;
}

function withPersistence < TState, TAction extends Action = AnyAction >(
    reducer: Reducer< TState, TAction >,
    serialize: ( state: TState ) => JSON,
    deserialize: ( json: JSON ) => TState,
): SerializableReducer< TState, TAction > {
    reducer.serialize = serialize;
    reducer.deserialize = deserialize;
    return reducer;
}
```

And then all reducers that previously declared serialization strategies should be wrapped with the new strategy... the problem, however, is that we can currently dispatch a single action handled by many reducers. With the new approach we'd have to traverse the reducer tree and call serialize on all of them... I'm not seeing how this would be easy, but I'm not super familiar with how all this works. Do you have any more details that might help @jsnajdr or @sgomes?

Reply:
> With the new approach we'd have to traverse the reducer tree and call serialize on all of them...

@saramarcondes All we need to make it work seamlessly is to patch the `combineReducers` function to also combine the children's `{de}serialize` methods. Example of how it would be used in the end:
```js
// reducer with no persistence
function user( state, action ) {
  switch ( action.type ) { ... }
}

// reducer with persistence
function posts( state, action ) {
  switch ( action.type ) { ... }
}

posts.deserialize = ( json: any ): TState => {
  if ( state === undefined ) {
    return getInitialState( posts );
  }
  if ( ! validateWithSchema( json, postsSchema ) ) {
    return getInitialState( posts );
  }
  return json as TState;
}

posts.serialize = ( state: TState ): any => {
  return state;
}

// the combined app reducer
const appReducer = combineReducers( {
  user,
  posts,
} );

// the combined app reducer can serialize/deserialize, too
const booState = appReducer.deserialize( jsonFromStorage );
storage.store( 'state', appReducer.serialize( state );
```

The `user` reducer doesn't define any persistence. That means that its state should not be stored at all. In other words, persistence is opt-in. `combinedReducer.serialize` should return an object that doesn't have an `user` key. And `combinedReducer.deserialize` should return the `user` reducer's initial state.

The `posts` reducer defines a de-serialization that validates the JSON with schema. That code can be extracted into a new version of the `withSchemaValidation` helper. Other than that, the serialization/deserialization is an identity, which stores the state verbatim without any transforms.